### PR TITLE
Feature/tag system

### PR DESCRIPTION
**Pull request by zaclegarssure**
Add a tag system with a fragment view to see and optionally (if authorized) edit them.

It tries to follow android mvvm design pattern and therefore may seems to contain a lot of boilerplate code.

## What?
Add a tag system, with one fragment, [TagsDisplayFragment.kt](https://github.com/PalFinderTeam/palFinder/pull/48/files#diff-eb14fbba6a2e449a49fd350c354d8d4a09db267d3be3242b13413448bcfdb3d3), to view and edit (if authorized) a set of tags, and a second one, [TagSelectorFragment.kt](https://github.com/PalFinderTeam/palFinder/pull/48/files#diff-4ce8eb58d671c94fb8d61375e20cef83a40483e3ded071a2b0aa1d72d4b30d8b), to select some missing tags and add them all together.
Overall it tries to follow android's mvvm [design pattern](https://developer.android.com/jetpack/guide) but is maybe a bit overkill here.

## How?
There are two example activities provided in the `tag/example` package to show how it works.
The overall design is as follows:

- `tag/TagsViewModel.kt` Provide an nice interface to get some tags, and if possible add and remove them.
- `tag/TagsViewModelFactory.kt` Provide just a function to build a TagsViewModel, because you can't easily pass arguments to a viewModel constructor, you need to use a factory object. Here it is just to pass the TagsRepository to the viewModel.
- `tag/TagsRepository` This is the interface that the real source of the tag needs to provide, the idea is that tags may come from a User object, or an Activity one and so on and providing a single interface help to abstract this. This file also include two simple implementation, which are useful to test the system if you don't have a tag source.
- `tag/TagAdapter` The tagSelector fragment uses a RecyclerView to display a list of selectable tags, and it requires an adapter class. It is also the class that specifies how the tag are filtered when searching.
- `tag/TagDisplayFragment.kt` `tag/TagSelectorFragment.kt` Those are the two fragment, they recover the viewModel implicitly from the activity that holds them.
- `Tag.kt` `Category.kt` The tags system is generic over the tag set itself, `Tag.kt` is the interface while `Category.kt` is a concrete implementation that will later contains all the tags we use for the activities, now it is filled with dummy tags. Being generic helps to create test tag sets.

## Testing?
I've added some unit tests for the viewModel, and it works, except that they don't appear in jacoco reports because of a bug. Otherwise it is really hard (maybe impossible actually) to inject a viewModel inside a fragment at test Time without relying on some custom setters, or a framework like Hilt or Koin. I therefore give up and we will wait that those fragments are in some real activities to perform some tests.

### CODE REVIEW

**Comment by TruddyTheDuddi** [comment]
So far it looks like a very solid and well built system, nice job

File: app/src/main/java/com/github/palFinderTeam/palfinder/tag/TagsDisplayFragment.kt [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+        val fm = parentFragmentManager
+        val currentTags = viewModel.tagContainer.value
+        val allTags = viewModel.allTags
+
+        if (currentTags != null) {
+            // Only show tags that are not already added
+            val tagsOptions = allTags.toSet().minus(currentTags)
+
+            TagSelectorFragment(tagsOptions.toList()) { tag -> viewModel.addTag(tag)}
+                .show(fm, "Tag selector")
```
Comment (TheblueMan003) [inline app/src/main/java/com/github/palFinderTeam/palfinder/tag/TagsDisplayFragment.kt:78]: Hard-coded string "Tag selector". You should use resources.

**Review by TheblueMan003 — APPROVED** [review]
Might need a function to go from NonEditableTags to EditableTags and vis-versa
They would also need to implements the Serializable interface to be able to pass them inside Intents
Otherwise, everything else looks amazing!

**Review by TruddyTheDuddi — APPROVED** [review]
Code is clean, other than the remark on the serializable interface inheritance by @TheblueMan003, the rest seems solid to me

**Pull request merged**